lenskit-slim/src/main/java/org/lenskit/slim/LinearRegressionHelper.java

```diff
@@ -0,0 +1,143 @@
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as
+ * published by the Free Software Foundation; either version 2.1 of the
+ * License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE. See the GNU General Public License for more
+ * details.
+ *
+ * You should have received a copy of the GNU General Public License along with
+ * this program; if not, write to the Free Software Foundation, Inc., 51
+ * Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
+ */
+package org.lenskit.slim;
+
+import it.unimi.dsi.fastutil.longs.*;
+import org.lenskit.util.math.Scalars;
+import java.util.Map;
+import static java.lang.Math.abs;
+
+
+/**
+ * Several simple vector algebra used by SLIM learning process
+ * @author <a href="http://www.grouplens.org">GroupLens Research</a>
+ */
+public final class LinearRegressionHelper {
+    /**
+     * soft-Thresholding for coordinate descent
+     * @param z
+     * @param gamma
+     * @return
+     */
+    public static double softThresholding(double z, double gamma) {
+        if (z > 0 && gamma < abs(z)) return z - gamma;
+        if (z < 0 && gamma < abs(z)) return z + gamma; // comment this line to get non-negative weights during learning process
+        return 0.0;
+    }
+
+    /**
+     * Transpose of matrix
+     * Convert a row view of rating matrix to a column view and vice versa
+     * After conversion, the key of empty row/column in the input matrix will be eliminated in the transpose matrix.
+     * So calling transposeMap(transposeMap(X)).keySet() may not equal to X.keySet() unless X has no mapping of key to empty Long2Double map.
+     *
+     * @param matrix
+     * @return
+     */
+    public static Long2ObjectMap<Long2DoubleMap> transposeMap(Long2ObjectMap<Long2DoubleMap> matrix) {
+        Long2ObjectMap<Long2DoubleMap> mapT = new Long2ObjectOpenHashMap<>();
+
+        for (Map.Entry<Long, Long2DoubleMap> rowEntry : matrix.entrySet()) {
+            long rowNum = rowEntry.getKey();
+            Long2DoubleMap rowValue = rowEntry.getValue();
+            for (Map.Entry<Long, Double> entry : rowValue.entrySet()) {
+                long colNum = entry.getKey();
+                double Value = entry.getValue();
+                Long2DoubleMap rowOfMapT = mapT.get(colNum);
+                if (rowOfMapT == null) rowOfMapT = new Long2DoubleOpenHashMap();
+                rowOfMapT.put(rowNum, Value);
+                mapT.put(colNum, rowOfMapT);
+            }
+        }
+        return mapT;
+    }
+
+    /**
+     * add two vectors
+     * @param a
+     * @param b
+     * @return
+     */
+    public static Long2DoubleMap addVectors(Long2DoubleMap a, Long2DoubleMap b) {
```

Comment:
We have a similar version 'Vectors.combine'. But 'Vectors.combine' may only combine the elements in common positions between two vectors. It may not work correctly for the sparse vector addition.